dbms/src/Functions/FunctionsString.cpp

```diff
@@ -5022,52 +5027,410 @@ class FunctionPosition : public IFunction
 
-        auto col_res = ColumnInt64::create();
-        int val_num = c0_col->size();
-        col_res->reserve(val_num);
+        block.getByPosition(result).column = std::move(col_res);
+    }
+
+    static void vectorVector(
+        const ColumnString::Chars_t & col0_data,
+        const ColumnString::Offsets & col0_offsets,
+        const ColumnString::Chars_t & col1_data,
+        const ColumnString::Offsets & col1_offsets,
+        PaddedPODArray<Int64> & res)
+    {
+        size_t pos;
+        size_t row_num = col0_offsets.size();
+        ColumnString::Offset prev_col0_str_offset = 0;
+        ColumnString::Offset prev_col1_str_offset = 0;
 
-        for (int i = 0; i < val_num; i++)
+        for (size_t i = 0; i < row_num; i++)
         {
-            c0_col->get(i, c0_field);
-            c1_col->get(i, c1_field);
+            size_t col0_str_len = col0_offsets[i] - prev_col0_str_offset - 1;
+            size_t col1_str_len = col1_offsets[i] - prev_col1_str_offset - 1;
+
+            if unlikely (col0_str_len == 0)
+            {
+                res[i] = 1;
+            }
+            else
+            {
+                LibCASCIICaseSensitiveStringSearcher searcher = LibCASCIICaseSensitiveStringSearcher(
+                    reinterpret_cast<const char *>(&col0_data[prev_col0_str_offset]),
+                    col0_str_len);
 
-            String c0_str = c0_field.get<String>();
-            String c1_str = c1_field.get<String>();
+                pos = searcher.search(&col1_data[prev_col1_str_offset], &col1_data[col1_offsets[i] - 1])
+                    - &col1_data[prev_col1_str_offset];
 
-            // return -1 when c1_str not contains the c0_str
-            Int64 idx = c1_str.find(c0_str);
-            col_res->insert(getPositionUTF8(c1_str, idx));
+                if (pos != col1_str_len)
+                    res[i] = 1 + pos;
+                else
+                    res[i] = 0;
+            }
+
+            prev_col0_str_offset = col0_offsets[i];
+            prev_col1_str_offset = col1_offsets[i];
         }
+    }
 
-        block.getByPosition(result).column = std::move(col_res);
+    static void vectorConst(
+        const ColumnString::Chars_t & col0_data,
+        const ColumnString::Offsets & col0_offsets,
+        const String & col1_str,
+        PaddedPODArray<Int64> & res)
+    {
+        size_t pos;
+        size_t row_num = col0_offsets.size();
+        ColumnString::Offset prev_col0_str_offset = 0;
+        size_t col1_str_len = col1_str.size();
+
+        for (size_t i = 0; i < row_num; i++)
+        {
+            size_t col0_str_len = col0_offsets[i] - prev_col0_str_offset - 1;
+
+            if unlikely (col0_str_len == 0)
+            {
+                res[i] = 1;
+            }
+            else
+            {
+                LibCASCIICaseSensitiveStringSearcher searcher = LibCASCIICaseSensitiveStringSearcher(
+                    reinterpret_cast<const char *>(&col0_data[prev_col0_str_offset]),
```

Comment:
Can we check the signature of this construction? It seems accepts "needle" instead of "haystack" as arguement.

Reply:
Confirmed that locate(needle, haystack), that's all right.